apps/oc_chef_wm/src/oc_chef_wm_keys.erl

```diff
@@ -0,0 +1,129 @@
+                           post_is_create/2,
+                           forbidden/2,
+                           is_authorized/2,
+                           service_available/2]}]).
+
+
+%% chef_wm behaviour callbacks
+-behaviour(chef_wm).
+-export([auth_info/2,
+         init/1,
+         init_resource_state/1,
+         malformed_request_message/3,
+         request_type/0,
+         validate_request/3,
+         to_json/2 ]).
+
+-export([allowed_methods/2,
+         create_path/2,
+         conflict_message/1,
+         from_json/2]).
+
+init(Config) ->
+    oc_chef_wm_base:init(?MODULE, Config).
+
+init_resource_state(_Config) ->
+    {ok, #keys_state{}}.
+
+request_type() ->
+    "keys".
+
+allowed_methods(Req, State) ->
+    {['GET'], Req, State}.
+
+validate_request('GET', Req, #base_state{resource_args = ObjectType, chef_db_context = Ctx, organization_guid = OrgId} = State) ->
+    ObjectName = chef_wm_util:object_name(ObjectType, Req),
+    ResourceState = make_resource_state_for_object(Ctx, ObjectType, ObjectName, OrgId),
+    {Req, State#base_state{resource_state = ResourceState }}.
+
+make_resource_state_for_object(DbContext, client, Name, OrgId) ->
+    make_resource_state_for_object(client, chef_db:fetch(#chef_client{name = Name, org_id = OrgId}, DbContext));
+make_resource_state_for_object(DbContext, user, Name, _OrgId) ->
+    make_resource_state_for_object(user, chef_db:fetch(#chef_user{username = Name}, DbContext)).
+
+make_resource_state_for_object(_Type, not_found) ->
+    #keys_state{};
+make_resource_state_for_object(Type, Object) ->
+    #keys_state{type = Type, name = chef_object:name(Object), authz_id = chef_object:authz_id(Object), id = chef_object:id(Object)}.
+
+
+%% Permissions:
+%% The permissios model here and in named_key will be the same
+%% Keys are considered an attribute of the object that owns them:
+%%   * target object (user or client) must exist
+%%   * In order to view a target's keys, requestor will need read access to that target. (handled in default is_authorized)
+%%   * In order to update/delete/add a key for a target, requestor will need update access to that target.
+%%   * If requestor is a user and target is a client, user must in the same org as the client
+%%   * If requestor is a user and target is self, requestor gets full permissions without specific acl checks.
+%%
+%% Before proceeding to authorize, first ensure that the owning actor (client or user) specified in the url exists.
+auth_info(Req, #base_state{resource_args = TargetType, resource_state = #keys_state{id = undefined}} = State) ->
+    Name = chef_wm_util:object_name(TargetType, Req),
+    Message = chef_wm_util:not_found_message(TargetType, Name),
+    Req1 = chef_wm_util:set_json_body(Req, Message),
+    {{halt, 404}, Req1, State#base_state{log_msg = {TargetType, not_found}}};
+%% If requestor is a user and target is self, requestor is authorized
+auth_info(Req, #base_state{ requestor_id = RequestorAuthzId,
```

Comment:
We've historically done it because sometimes people can do bad things to their permissions, but we don't want them every to break access to themselves. 

I think in the default case, it will work fine without special casing - but if someone accidentally removes the wrong ACE...